Work on minefield branch

Qsionc opened the pull request:
My early work on minefield branch. 
Some brutal code. 

Main changes to minefield.rs in methods that place mines (placing hints to be done)
Also, flood_uncover method is implemented, not tested, since graphics and parts of logic still missing.

Qsionc committed:
feat: start implementing enimdnal logic

implement TODO methods with my proposal
place_hints methods in TODO

future work: graphics module and finish minefield

tranzystorekk commented on src/minefield.rs:
```diff
@@ -55,8 +58,10 @@ pub struct Tile {
 #[derive(Debug)]
 pub struct Board {
     tiles: Vec<Tile>,
+    covered_count: usize,
```
bikeshed: `covered` should be sufficient

Qsionc replied:
Done.

tranzystorekk commented on src/minefield.rs:
```diff
@@ -144,15 +159,36 @@ impl Board {
-        todo!()
+        let tile_idx = Self::coords_to_index(&self.params, x, y);
+        match self.tiles[tile_idx].cover {
+            Cover::Up(_) => {
```
This uncovers even flagged tiles, if I'm not mistaken

Qsionc replied:
Done. Please review.

tranzystorekk replied:
For now, we can handle cover state much simpler with a

```rust
if !tile.is_uncoverable() {
    return;
}

tranzystorekk commented on src/minefield.rs:
```diff
@@ -81,6 +86,10 @@ impl Tile {
     }
 
+    pub fn uncover(&mut self) {
+        self.cover = Cover::Down;
```
I don't think a method for a single assignment is necessary

Qsionc replied:
Done.

tranzystorekk commented on src/minefield.rs:
```diff
@@ -163,22 +199,77 @@ impl Board {
+    /// Get single dimension index of 2D tile
+    ///
+    /// This method exists purely because borrow checker took me hostage :)
+    fn coords_to_index(params: &Params, x: usize, y: usize) -> usize {
```
Not sure how the `&self` borrow could possibly be failing, do you have an example?

Qsionc replied:
It is fixed now :)

tranzystorekk commented on src/minefield.rs:
```diff
@@ -163,22 +199,77 @@ impl Board {
-        todo!()
+        // i would put (usize, usize) here, since its just one point user clicks on + eventual flood
+        // and then have this method be called in handle_uncover at the beginning
+        if self.placed {
```
I'd rather we placed such checks at call site, e.g. in `handle_uncover`

Qsionc replied:
Done. Please look into new commit. Check is now performed in `handle_uncover`

tranzystorekk commented on src/minefield.rs:
```diff
@@ -163,22 +199,77 @@ impl Board {
+            self.tiles[*index].object = Object::Mine;
+        }
+
+        self.place_hints();
```
Poor separation of concerns, it would be better to create a single focus point for the deferred setup in a new method, e.g. `place_mines_and_hints`, that calls both `place_mines` and `place_hints`

Qsionc replied:
Done

tranzystorekk commented on src/minefield.rs:
```diff
@@ -163,22 +199,77 @@ impl Board {
+
+        while let Some(index) = set.iter().next() {
+            self.tiles[*index].object = Object::Mine;
+        }
```
I suggest you take a look at https://docs.rs/rand/latest/rand/seq/trait.IteratorRandom.html,

we can just generate a range of all possible indices (excluding `skip`) and make rand choose however many mines we need.

Qsionc replied:
Done

tranzystorekk commented on src/minefield.rs:
```diff
@@ -163,22 +199,77 @@ impl Board {
     fn flood_uncover(&mut self, x: usize, y: usize) {
-        todo!()
+        let all_tiles = &mut self.tiles;
+        let mut tile_pos = VecDeque::new();
```
We don't need the queue (aka BFS) variant, so a simple `Vec` + `push` + `pop` should be good

Qsionc replied:
To be done

tranzystorekk commented on src/minefield.rs:
```diff
@@ -163,22 +199,77 @@ impl Board {
+
+        while let Some(tile) = tile_pos.pop_front() {
+            let t_idx = Self::coords_to_index(&self.params, tile.0, tile.1);
+            if all_tiles[t_idx].is_uncoverable() && !all_tiles[t_idx].is_mine() {
```
`all_tiles[...]` is repeated needlessly a lot here, we can just do `let tile = &mut self.tiles[...];` instead

Qsionc replied:
Does not work. :)

tranzystorekk commented on src/minefield.rs:
```diff
@@ -163,22 +199,77 @@ impl Board {
+            let t_idx = Self::coords_to_index(&self.params, tile.0, tile.1);
+            if all_tiles[t_idx].is_uncoverable() && !all_tiles[t_idx].is_mine() {
+                all_tiles[t_idx].uncover();
+                if self.covered_count > self.params.mines {
```
Not sure if this check is needed, seems overly defensive

Qsionc replied:
Leaving it for now, i like being defensive. :)

tranzystorekk commented on src/minefield.rs:
```diff
@@ -144,15 +159,36 @@ impl Board {
     pub fn handle_mark(&mut self, x: usize, y: usize) {
-        todo!()
+        let tile_idx = Self::coords_to_index(&self.params, x, y);
+        if let Cover::Up(mark) = self.tiles[tile_idx].cover {
```
This copies `mark` instead of getting a mutable reference to it

```suggestion
        if let Cover::Up(mark) = &mut self.tiles[tile_idx].cover {
```

Qsionc replied:
Yet to be implemented

tranzystorekk commented on src/minefield.rs:
```diff
@@ -144,15 +159,36 @@ impl Board {
-        todo!()
+        let tile_idx = Self::coords_to_index(&self.params, x, y);
+        if let Cover::Up(mark) = self.tiles[tile_idx].cover {
+            match mark {
+                Mark::Flag => self.tiles[tile_idx].cover = Cover::Up(Mark::Unsure),
+                Mark::Unsure => self.tiles[tile_idx].cover = Cover::Up(Mark::None),
+                Mark::None => self.tiles[tile_idx].cover = Cover::Up(Mark::Flag),
+            }
```
Use `Mark::cycle` instead

Qsionc replied:
Not yet implemented

tranzystorekk commented on src/minefield.rs:
```diff
@@ -144,15 +159,36 @@ impl Board {
+                    _ => ()
+                }
+            }
+            Cover::Down => () //OPTIONAL: on uncover additional action when clicking hint
```
We'll add this as a separate feature after initial playable version

tranzystorekk commented on src/minefield.rs:
```diff
@@ -163,22 +199,77 @@ impl Board {
+                    let max_y = if tile.1 < self.params.height { tile.1 + 1 } else { tile.1 };
+
+                    for i in min_x..=max_x {
+                        for j in min_y..=max_y {
```
We could use a `fn neighbors(&self, x: usize, y: usize) -> impl Iterator<Item = (usize, usize)>` method

Qsionc replied:
Not yet implemented. This knowledge eludes me.

tranzystorekk commented on src/minefield.rs:
```diff
@@ -121,16 +136,16 @@ impl Board {
 
     pub fn is_victory(&self) -> bool {
-        todo!()
+        !self.defeat && self.covered_count == self.params.mines
```
The `!self.defeat` check seems overly defensive, victory and defeat are mutually exclusive conditions

Qsionc replied:
Not touched yet

Qsionc committed:
feat(improvement): Work on minefield branch part 2

Minor improvements:
1. Used IteratorRandom to generate mine placements
2. Split responsibilities. Now handle_uncover takes care
of mine and hint placement if it is not done yet
3. Overall code minor improvements

not all comments were taken into account in this commit.

tranzystorekk commented on src/minefield.rs:
```diff
@@ -175,10 +248,23 @@ impl Board {
-        todo!()
+        // i would put (usize, usize) here, since its just one point user clicks on + eventual flood
+        // and then have this method be called in handle_uncover at the beginning
+        self.placed = true;
```
Move this to handle_uncover as well

Qsionc replied:
Done

tranzystorekk commented on src/minefield.rs:
```diff
@@ -175,10 +248,23 @@ impl Board {
+        // i would put (usize, usize) here, since its just one point user clicks on + eventual flood
+        // and then have this method be called in handle_uncover at the beginning
+        self.placed = true;
+        let mut rng = thread_rng();
+        let idx_range = Range { start: 0, end: self.params.height * self.params.height };
+
+        for i in 0..self.params.mines {
+            while let Some(mine_idx) = idx_range.clone().choose(&mut rng) {
+                if skip.contains(&mine_idx) {
+                    continue;
+                }
+                self.tiles[mine_idx].object = Object::Mine;
+            }
+        }
```
I was thinking about something more like:

```rust
(0..self.tiles.len()).filter(|i| !skip.contains(i)).choose_multiple(&mut rng, self.params.mines)
```

tranzystorekk replied:
(Also, this particular implementation might choose the same index multiple times and then we end up with fewer mines than required)

Qsionc replied:
Done. Used your example implementation with a twist.

tranzystorekk commented on src/minefield.rs:
```diff
@@ -163,9 +206,39 @@ impl Board {
 
+    /// Get single dimension index of 2D tile
+    ///
+    /// This method exists purely because borrow checker took me hostage :)
```
This comment doesn't seem necessary anymore :P

Qsionc replied:
:) Done

Qsionc committed:
feat: Work on minefield branch part 3

Implement hint placement
Implement neighbours method
Implement new way of handling flood-uncover
Minor refactoring

tranzystorekk commented on src/minefield.rs:
```diff
@@ -144,15 +158,68 @@ impl Board {
+            (1, 0),
+            (1, 1),
+        ];
+        let width = self.params.width;
```
Please leave a comment here that says why this is necessary, e.g.

```rust
// copy these values to avoid borrowing `self` in the closure
```

tranzystorekk commented on src/minefield.rs:
```diff
@@ -144,15 +158,68 @@ impl Board {
+            let new_y = (y as i32 + off_y);
+            let new_x_inbounds = new_x >= 0 && new_x < width as i32;
+            let new_y_inbounds = new_y >= 0 && new_y < height as i32;
+            if new_x_inbounds && new_y_inbounds {
```
Invert this so that happy path stays at the end of the closure

tranzystorekk commented on src/minefield.rs:
```diff
@@ -175,10 +266,38 @@ impl Board {
     fn place_mines(&mut self, skip: &[usize]) {
-        todo!()
+        // i would put (usize, usize) here, since its just one point user clicks on + eventual flood
+        // and then have this method be called in handle_uncover at the beginning
```
These comments are probably unhelpful

tranzystorekk commented on src/minefield.rs:
```diff
@@ -175,10 +266,38 @@ impl Board {
+                }
+                let mine_count = self
+                    .neighbours(x, y)
+                    .filter(|pos| self.tiles[self.coords_to_index(pos.0, pos.1)].is_mine())
```
`pos` can be destructured into `(x, y)`

tranzystorekk commented on src/minefield.rs:
```diff
@@ -175,10 +266,38 @@ impl Board {
+            .filter(|i| !skip.contains(i))
+            .choose_multiple(&mut rng, self.params.mines);
+
+        for mine in &mines {
```
The `&mines` borrow is unnecessary

Qsionc committed:
feat: Work on minefield branch part 4

Do minor refactoring to comments

tranzystorekk commented on src/minefield.rs:
```diff
@@ -144,15 +157,65 @@ impl Board {
+
+        if !self.placed {
+            self.placed = true;
+            self.place_mines(&[tile_idx]);
```
This should also include the neighboring tiles

Qsionc replied:
What do you mean by that? To ensure that there is a square around it?

tranzystorekk replied:
yup, otherwise you can get cornered on the first move 😆

Qsionc replied:
New patch with improvement regarding this comment :)

Qsionc committed:
feat: Work on minefield branch part 4.5

Improved place mines algorithm - now includes first click neighbours
to make sure one does not get stuck surrounded by mines.
